date/ls: Switch from chrono to jiff

drinkcat opened the pull request:
`jiff` handling of timezones makes our life so much easier. Also, jiff provides
a convenient feature to only embed the timezone database when necessary (e.g. Windows).

 - Fixes #7497, #7498, #7659.
 - Helps with #7504, but still adds an extra ~300k compared to a chrono-only solution
   (that did not fix all issues, FWIW):
```
-rwxr-xr-x 2 drinkcat drinkcat 4828144 Apr 27 13:57 target/release/date ## HEAD
-rwxr-xr-x 2 drinkcat drinkcat 3279432 Apr 28 15:42 target/release/date ## jiff
-rwxr-xr-x 2 drinkcat drinkcat 2949080 Apr 27 13:58 target/release/date ## with #7849
```

Part of the tests taken from #7597 (@jadijadi FYI), then refactored on top of those to add more test cases.

@BurntSushi thanks for your support!

--

### Cross.toml: Install tzdata in container

Linux tests require that now, as we now assume /usr/share/zoneinfo
is present.

### test_date: Extend coverage to a lot more timezones

Also test %z/%Z formats.

### date: Add more TZ tests

[drinkcat: separated test changes]

### ls: switch to lenient formating configuration

### ls: Avoid additional String creation/copy in display_date

From code provided in #7852 by @BurntSushi.

Depending on the benchmarks, there is _still_ a small performance
difference (~4%) vs main, but it's seen mostly on small trees
getting printed repeatedly, which is probably not a terribly
interesting use case.

### ls: cache recent time threshold in jiff implementation

### ls: convert to jiff

### Revert "ls: Optimize time formatting"

This reverts commit fc6b896c271eed5654418acc267eb21377c55690.

This also reverts the change from new to new_lenient, we'll
recover that later as part of the jiff conversion.

### date: switch from chrono to jiff

Also adds cargo dependency.

sylvestre commented on src/uu/date/src/date.rs:
```diff
@@ -403,27 +390,51 @@ fn make_format_string(settings: &Settings) -> &str {
+    match parse_datetime::parse_datetime(s.as_ref()) {
+        Ok(date) => {
+            let timestamp =
+                Timestamp::new(date.timestamp(), date.timestamp_subsec_nanos() as i32).unwrap();
```
for once, it might make sense ?! :)

drinkcat replied:
Ah, yeah, it does make some sort of sense... But:

The suggestion doesn't build:
> `?` couldn't convert the error to `(std::string::String, ParseDateTimeError)`

The error cannot really happen anyway:
> This returns an error if the given components would correspond to an instant outside the supported range. Also, nanosecond is limited to the range -999,999,999..=999,999,999.

So yes, technically I could handle the error and convert it, but since `timestamp()` and `timestamp_subsec_nanos()` both come from a good library (jiff), and go into another good library (chrono), I think we can just `unwrap`.

And we can get rid of this conversion once we switch `parse_datetime` to jiff (I haven't found time to look into that yet...).

drinkcat commented:
Rebased, and also cleaned up `custom_tz_fmt` that's not needed anymore.

I'll try to continue this conversion (maybe will look at parse_datetime next).

drinkcat commented on src/uu/ls/src/ls.rs:
```diff
@@ -274,64 +273,41 @@ enum TimeStyle {
+        let mut out = StdIoWrite(out);
+        let config = jiff::fmt::strtime::Config::new().lenient(true);
+
+        match (self, recent) {
+            (Self::FullIso, _) => {
+                tm.format_with_config(&config, "%Y-%m-%d %H:%M:%S.%f %z", &mut out)
             }
+            (Self::LongIso, _) => tm.format_with_config(&config, "%Y-%m-%d %H:%M", &mut out),
+            (Self::Iso, true) => tm.format_with_config(&config, "%m-%d %H:%M", &mut out),
+            (Self::Iso, false) => tm.format_with_config(&config, "%Y-%m-%d ", &mut out),
+            // spell-checker:ignore (word) datetime
+            //In this version of chrono translating can be done
+            //The function is chrono::datetime::DateTime::format_localized
+            //However it's currently still hard to get the current pure-rust-locale
+            //So it's not yet implemented
+            (Self::Locale, true) => tm.format_with_config(&config, "%b %e %H:%M", &mut out),
+            (Self::Locale, false) => tm.format_with_config(&config, "%b %e  %Y", &mut out),
+            (Self::Format(fmt), _) => tm.format_with_config(&config, fmt, &mut out),
         }
```
That's better yes! Thanks!

cakebaker commented on tests/by-util/test_date.rs:
```diff
@@ -564,11 +566,121 @@ fn test_date_from_stdin() {
+#[test]
+fn test_date_utc_time() {
+    // Test that -u flag shows correct UTC time
+    new_ucmd!().arg("-u").arg("+%H:%M").succeeds();
```
Clever approach!

cakebaker commented:
Good work, thanks!

sylvestre commented:
+1
